pkg/endpoint/policy.go

```diff
@@ -422,22 +422,21 @@ func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMe
 			return
 		}
 		e.RUnlock()
-		scopedLog := e.getLogger()
 
 		// We should only queue the request after we use all the endpoint's
 		// lock/unlock. Otherwise this can get a deadlock if the endpoint is
 		// being deleted at the same time. More info PR-1777.
 		doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
 		if doneFunc != nil {
-			scopedLog.Debug("Dequeued endpoint from build queue")
+			e.getLogger().Debug("Dequeued endpoint from build queue")
```

Comment:
I think at least before and after `regenerate()`, the scopedLog must be refreshed as it will have changed significantly. We should optimize for correctness in the logs instead of micro-optimizing this code at this level. If we want to optimize then let's ensure that we don't call `getLogger()` when debugging is not enabled.

Reply:
Nit-picking here, but `getLogger()` is called on average about 1 time only. On monitorAPI failure it's 2 times, but there is no need to optimize that.

Reply:
The reason I did this is because there is a lot of change in state to the endpoint during this process, and I found that information about the endpoint was stale in the log (e.g., endpoint ID was zero at certain points in time even after it had been added to the endpoint manager and allocated an ID). I think the only time we can safely use a scoped log is if the endpoint is locked, and the endpoint structure isn't locked here.